### PR TITLE
Comments between assignment (without let) lhs and rhs handling

### PR DESCRIPTION
**Pull request by davidBar-On**
Add handling of comments between assignment (without let) lhs and rhs by using `rewrite_assign_rhs_with_comments()` (one of the cases identified in [#4626 discussion](https://github.com/rust-lang/rustfmt/pull/4626#issuecomment-761817663)).

### CODE REVIEW

File: src/formatting/expr.rs [hunk 1]
```diff
@@ -1895,16 +1895,22 @@ fn rewrite_assignment(
...
-        Some(op) => context.snippet(op.span),
-        None => "=",
+    let span_hi = rhs.span.lo();
+    let (operator_str, span_lo) = match op {
+        Some(op) => (context.snippet(op.span), op.span.hi()),
+        None => {
+            let lo = lhs.span.hi();
+            let offset = context.snippet(mk_sp(lo, span_hi)).find_uncommented("=")? + 1;
+            ("=", lo + BytePos(offset as u32))
+        }
```
Comment (calebcartwright) [inline src/formatting/expr.rs:1905]: I could be remembering incorrectly, but I believe the AST nodes for assignment expressions actually contain the span for the operator or binop. Could you double check that (presumably back at the call sites), and if it's there then update our destructuring code to grab it?

Would much rather plug and play from the AST than recalculating spans when possible

File: src/formatting/expr.rs [hunk 2]
```diff
@@ -1930,6 +1936,26 @@ pub(crate) fn rewrite_assign_rhs<S: Into<String>, R: Rewrite>(
...
+        context,
+        lhs,
+        ex,
+        shape,
+        RhsTactics::Default,
+        between_span,
+        true,
+    )
+}
+
```
Comment (calebcartwright) [inline src/formatting/expr.rs:1958]: I'm okay with adding another helper function here to avoid repetitive specification of the rhs tactics, but I think we'd probably want to expose as a parameter the flag which controls whether comments are allowed between the assignment op and rhs without forcing line wrapping (i.e. the last arg in the `rewrite_assign_rhs_with_comments` as a param on `rewrite_assign_rhs_with_span` instead of a hardcoded `true`)

File: tests/source/lhs-to-rhs-between-comments/assignment.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    fn main () {
+    let var = /* Block comment longgggggggggggggggggggggggggggggggggggggggggggggggggggggggg */ first;
+    var /= /* Block comment longgggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg */ second;
+    }
+    fn main () {
+    let var = /* Block comment longgggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg */ first;
+    var /= /* Block comment longgggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg */ second;
+    }
+
```
Comment (calebcartwright) [inline tests/source/lhs-to-rhs-between-comments/assignment.rs:48]: Solid test suite! For the sake of extreme thoroughness, do you think it'd be worthwhile to add a multiline blockstyle comment case?

**Review by calebcartwright — COMMENTED** [review]
Thanks for this! Looks pretty good, few minor comments left inline

**Comment by calebcartwright** [comment]
Also be sure to rebase to pull in the rustc-ap crate updates

File: src/formatting/expr.rs (hunk 1 above)
Comment (davidBar-On) [reply src/formatting/expr.rs]: Yes, there is such span.  Changed the code to use it.

File: src/formatting/expr.rs (hunk 2 above)
Comment (davidBar-On) [reply src/formatting/expr.rs]: I removed the helper function and changed the code to read directly `rewrite_assign_rhs_with_comments` instead, as during the evaluation of the multi-lie comments issue I found that using this helper function makes the code less clear.

File: tests/source/lhs-to-rhs-between-comments/assignment.rs (hunk 3 above)
Comment (davidBar-On) [reply tests/source/lhs-to-rhs-between-comments/assignment.rs]: Added multiline comments test cases.

However, I think there is an **issue** with the multiline comments formatting by `combine_strs_with_missing_comments()` (not caused by this PR so I didn't try to fix it here).  An example of the issue is that the following formatted code:
```rust
let var = /* Block comment line 1
    * Block comment line 2 */
    first;
```
Should probably have been **either**:
```rust
let var = /* Block comment line 1
           * Block comment line 2 */
    first;
```
**or**:
 ```rust
let var = /* Block comment line 1
            * Block comment line 2 */  first;
```
**Is this really an issue?  If yes, which of the output versions is desired?** If this is an issue that should be fixed, I can try to do that.

Comment (calebcartwright) [reply tests/source/lhs-to-rhs-between-comments/assignment.rs]: > Is this really an issue?

it is _an_ issue yes. some of these comment edge cases can be so exhausting sometimes :smile: it would have to be the former or what we did in other similar scenarios 

e.g. something like this (hand typed to give a visual, indentation may be off)
```rust
let var = 
    /* Block comment line 1
     * Block comment line 2 */
    first;
```

I'd prefer the former in the spirit of maintaining developer defined comment association, as odd as such a comment would be in these types of local cases.

**Comment by calebcartwright** [comment]
I am going to go ahead and merge this in the spirit of moving forward, though want to make a note of how it would potentially be problematic in the 1.x case.

Currently, in the presence of such comments rustfmt will bail and leave the original formatting in place, but with this change rustfmt will of course reformat all such cases. The multi-line block style comment that opens on the same line as the operator would be problematic though as however unlikely that actually is in practice, we'd be emitting formatting that's off which the user would be unable to correct.

Only going to merge here because it's against the master branch and it's such an extreme edge case that we'll get sorted quickly anyway. Let's prioritize getting that edge case resolved though :+1:
